lib/lookdevXUsd/UsdMxVersionUpgrade.cpp

```diff
@@ -0,0 +1,1097 @@
+                auto normalInput = node.GetInput(_tokens->normal);
+                auto tangentInput = node.GetInput(_tokens->tangent);
+                auto bitangentInput = node.GetInput(_tokens->bitangent);
+                if ((normalInput || tangentInput) && !bitangentInput) {
+                    auto ngPrim = node.GetPrim().GetParent();
+                    auto crossNode = _createSiblingNode(node, _tokens->ND_crossproduct_vector3, "normalmap_cross");
+                    normalInput.GetAttr().FlattenTo(crossNode.GetPrim(), UsdShadeUtils::GetFullName(_tokens->in1, UsdShadeAttributeType::Input));
+                    tangentInput.GetAttr().FlattenTo(crossNode.GetPrim(), UsdShadeUtils::GetFullName(_tokens->in2, UsdShadeAttributeType::Input));
+
+                    auto normalizeNode = _createSiblingNode(node, _tokens->ND_normalize_vector3, "normalmap_cross_norm");
+                    normalizeNode.CreateInput(_tokens->in, kMaterialXToUsdType.at(_tokens->vector3)).ConnectToSource(crossNode.CreateOutput(_tokens->out, kMaterialXToUsdType.at(_tokens->vector3)));
+                    node.CreateInput(_tokens->bitangent, kMaterialXToUsdType.at(_tokens->vector3)).ConnectToSource(normalizeNode.CreateOutput(_tokens->out, kMaterialXToUsdType.at(_tokens->vector3)));
+                }
+
+                node.SetShaderId(_tokens->ND_normalmap_float);
+            }
+        }
+    }
+    // Delete unused nodes:    
+    auto editor = UsdNamespaceEditor(usdMaterial.GetPrim().GetStage());
+    for (auto node : unusedNodes) {
+        editor.DeletePrim(node.GetPrim());
+        if (editor.CanApplyEdits()){
+            editor.ApplyEdits();
+        } else {
+            TF_WARN("Failed to delete obsolete node '%s' after material upgrade. Please make sure the material layer is writable.",
+                    node.GetPath().GetText());
+        }
+    }
+
+    // Update the version attribute using the MaterialXConfigAPI schema:
+    auto configAPI = UsdMtlxMaterialXConfigAPI::Apply(usdMaterial.GetPrim());
+    auto versionAttr = configAPI.CreateConfigMtlxVersionAttr();
+    versionAttr.Set(_tokens->currentMxVersion.GetString());
+}
+
+} // namespace
+
+namespace LookdevXUsd::Version
+{
+
+std::optional<std::string> isLegacyShaderGraph(const Ufe::Path& materialElementPath)
+{
+    const auto materialPath = _getMaterialPath(materialElementPath);
+    if (materialPath.empty()) {
+        return std::nullopt;
+    }
+
+    const auto materialItem = Ufe::Hierarchy::createItem(materialPath);
+    auto materialPrim = UsdShadeMaterial(MayaUsdAPI::getPrimForUsdSceneItem(materialItem));
+    if (!materialPrim)
+    {
+        return std::nullopt;
+    }
+    return _isLegacyMaterial(materialPrim);
+}
+
+void UpgradeStage(const Ufe::Path& stagePath)
+{
+    auto stage = MayaUsdAPI::getStage(stagePath);
+    if (!stage)
+    {
+        return;
+    }
+
+    for (const auto& prim : stage->Traverse()) {
```

Comment:
When you upgrade you do not use the default predicate but this Traverse overload does, could it thus have different behavior?

Reply:
I was just comparing to the custom predicate around line 422 of the materialX upgrade code.

ttps://github.com/Autodesk/maya-usd/pull/4405/changes/BASE..c2e5870ed295c52b96db27be936bc88eca274e26#diff-2d29e4820816e0b3eb41ddd2334c0d9f8194ac989ed30eba2880ea70d237551cR422